src/caffe/test/test_caffe_main.cpp

```diff
@@ -17,18 +18,30 @@ int main(int argc, char** argv) {
   caffe::GlobalInit(&argc, &argv);
 #ifndef CPU_ONLY
   // Before starting testing, let's first print out a few cuda defice info.
-  int device;
-  cudaGetDeviceCount(&device);
-  cout << "Cuda number of devices: " << device << endl;
+  std::vector<int> devices;
+  int device_count;
+
+  cudaGetDeviceCount(&device_count);
+  cout << "Cuda number of devices: " << device_count << endl;
+
   if (argc > 1) {
     // Use the given device
-    device = atoi(argv[1]);
-    cudaSetDevice(device);
-    cout << "Setting to use device " << device << endl;
+    devices.push_back(atoi(argv[1]));
+    cudaSetDevice(devices[0]);
   } else if (CUDA_TEST_DEVICE >= 0) {
     // Use the device assigned in build configuration; but with a lower priority
-    device = CUDA_TEST_DEVICE;
+    devices.push_back(CUDA_TEST_DEVICE);
   }
+
+  if (devices.size() == 1) {
```

Comment:
The logic around here reads in a confusing way to me... e.g., doesn't `cudaSetDevice` get called twice if a device is given on the command line? And the conditional here isn't really about the number of devices, it's about whether either case above was invoked...

I think this would read better if the `else` below were joined to the `if` above, so we have just three mutually exclusive cases. And then I guess we can call `cudaSetDevice` on the first device always (what's the behavior of the third case now in that respect? Do we even need to call it at all? Why wasn't it called for the build option before?).